src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,569 @@
+        /// type parameters of the specified target callable (the "innermost resolutions"). This mapping may potentially be to
+        /// type parameters of other callables, which are then further concretized by subsequent resolutions.
+        /// Returns the constructed dictionary as out parameter. Returns true if the combination of the given resolutions is valid,
+        /// i.e. if there are no conflicting resolutions and type parameters of the target callable are uniquely resolved
+        /// to either a concrete type, a type parameter of another callable, or themselves.
+        /// Throws an ArgumentNullException if the given target is null.
+        /// NOTE: This routine prioritizes the verifications to ensure the correctness of the resolution over performance.
+        /// </summary>
+        public static bool TryCombineTypeResolutionsWithTarget(QsQualifiedName target, out TypeParameterResolutions combined, params TypeParameterResolutions[] resolutions)
+        {
+            if (target == null) throw new ArgumentNullException(nameof(target));
+            var success = TryCombineTypeResolutions(out combined, resolutions);
+            combined = combined.Where(kvp => kvp.Key.Item1.Equals(target)).ToImmutableDictionary();
+            return success;
+        }
+
+        public static bool TryCombineTypeResolutions(out TypeParameterResolutions combined, params TypeParameterResolutions[] resolutionDictionaries)
+        {
+            if (!resolutionDictionaries.Any())
+            {
+                combined = TypeParameterResolutions.Empty;
+                return true;
+            }
+
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static Tuple<QsQualifiedName, NonNullable<string>> AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+            static bool IsSelfResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && tp.Item.Origin.Equals(typeParam.Item1) && tp.Item.TypeName.Equals(typeParam.Item2);
+            static bool IsConstrictiveResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && typeParam.Item1.Equals(tp.Item.Origin) && typeParam.Item2.Value != tp.Item.TypeName.Value;
+
+            foreach (var resolutionDictionary in resolutionDictionaries)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
```

Comment:
Looks like this version still contains the bug I had introduced that type parameters contained in the Resolution are not properly picked up. A ToDo would be good to not forget about it (I already forgot about it until I looking at the code just now).